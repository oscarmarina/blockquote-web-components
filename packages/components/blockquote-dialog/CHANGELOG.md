# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [1.3.4](https://github.com/oscarmarina/blockquote-web-components/compare/@blockquote-web-components/blockquote-dialog@1.3.3...@blockquote-web-components/blockquote-dialog@1.3.4) (2024-12-30)

**Note:** Version bump only for package @blockquote-web-components/blockquote-dialog





## [1.3.3](https://github.com/oscarmarina/blockquote-web-components/compare/@blockquote-web-components/blockquote-dialog@1.3.2...@blockquote-web-components/blockquote-dialog@1.3.3) (2024-12-26)

**Note:** Version bump only for package @blockquote-web-components/blockquote-dialog





## [1.3.2](https://github.com/oscarmarina/blockquote-web-components/compare/@blockquote-web-components/blockquote-dialog@1.3.1...@blockquote-web-components/blockquote-dialog@1.3.2) (2024-12-06)

**Note:** Version bump only for package @blockquote-web-components/blockquote-dialog





## [1.3.1](https://github.com/oscarmarina/blockquote-web-components/compare/@blockquote-web-components/blockquote-dialog@1.3.0...@blockquote-web-components/blockquote-dialog@1.3.1) (2024-11-28)

**Note:** Version bump only for package @blockquote-web-components/blockquote-dialog





# [1.3.0](https://github.com/oscarmarina/blockquote-web-components/compare/@blockquote-web-components/blockquote-dialog@1.2.1...@blockquote-web-components/blockquote-dialog@1.3.0) (2024-11-24)


### Features

* **blockquote-dialog:** update utilities ([6537d11](https://github.com/oscarmarina/blockquote-web-components/commit/6537d118b5eada5a9c178f9efcf727e9c15675ea))





## [1.2.1](https://github.com/oscarmarina/blockquote-web-components/compare/@blockquote-web-components/blockquote-dialog@1.2.0...@blockquote-web-components/blockquote-dialog@1.2.1) (2024-11-11)

**Note:** Version bump only for package @blockquote-web-components/blockquote-dialog





# [1.2.0](https://github.com/oscarmarina/blockquote-web-components/compare/@blockquote-web-components/blockquote-dialog@1.1.14...@blockquote-web-components/blockquote-dialog@1.2.0) (2024-11-08)


### Features

* **blockquote-dialog:** update exports ([f5150a8](https://github.com/oscarmarina/blockquote-web-components/commit/f5150a8ff7e94cf554e90d2ff0e2173027084f93))





## [1.1.14](https://github.com/oscarmarina/blockquote-web-components/compare/@blockquote-web-components/blockquote-dialog@1.1.13...@blockquote-web-components/blockquote-dialog@1.1.14) (2024-11-08)

**Note:** Version bump only for package @blockquote-web-components/blockquote-dialog





## [1.1.13](https://github.com/oscarmarina/blockquote-web-components/compare/@blockquote-web-components/blockquote-dialog@1.1.12...@blockquote-web-components/blockquote-dialog@1.1.13) (2024-10-18)

**Note:** Version bump only for package @blockquote-web-components/blockquote-dialog





## [1.1.12](https://github.com/oscarmarina/blockquote-web-components/compare/@blockquote-web-components/blockquote-dialog@1.1.11...@blockquote-web-components/blockquote-dialog@1.1.12) (2024-10-17)

**Note:** Version bump only for package @blockquote-web-components/blockquote-dialog





## [1.1.11](https://github.com/oscarmarina/blockquote-web-components/compare/@blockquote-web-components/blockquote-dialog@1.1.10...@blockquote-web-components/blockquote-dialog@1.1.11) (2024-10-05)

**Note:** Version bump only for package @blockquote-web-components/blockquote-dialog





## [1.1.10](https://github.com/oscarmarina/blockquote-web-components/compare/@blockquote-web-components/blockquote-dialog@1.1.9...@blockquote-web-components/blockquote-dialog@1.1.10) (2024-09-13)

**Note:** Version bump only for package @blockquote-web-components/blockquote-dialog





## [1.1.9](https://github.com/oscarmarina/blockquote-web-components/compare/@blockquote-web-components/blockquote-dialog@1.1.8...@blockquote-web-components/blockquote-dialog@1.1.9) (2024-09-13)

**Note:** Version bump only for package @blockquote-web-components/blockquote-dialog





## [1.1.8](https://github.com/oscarmarina/blockquote-web-components/compare/@blockquote-web-components/blockquote-dialog@1.1.7...@blockquote-web-components/blockquote-dialog@1.1.8) (2024-09-06)

**Note:** Version bump only for package @blockquote-web-components/blockquote-dialog





## [1.1.7](https://github.com/oscarmarina/blockquote-web-components/compare/@blockquote-web-components/blockquote-dialog@1.1.6...@blockquote-web-components/blockquote-dialog@1.1.7) (2024-07-26)


### Bug Fixes

* **blockquote-dialog:** update event-listeners ([41612f6](https://github.com/oscarmarina/blockquote-web-components/commit/41612f63c0547aa29d628f5c6aec7319de58036c))





## [1.1.6](https://github.com/oscarmarina/blockquote-web-components/compare/@blockquote-web-components/blockquote-dialog@1.1.5...@blockquote-web-components/blockquote-dialog@1.1.6) (2024-07-26)

**Note:** Version bump only for package @blockquote-web-components/blockquote-dialog





## [1.1.5](https://github.com/oscarmarina/blockquote-web-components/compare/@blockquote-web-components/blockquote-dialog@1.1.4...@blockquote-web-components/blockquote-dialog@1.1.5) (2024-07-25)

**Note:** Version bump only for package @blockquote-web-components/blockquote-dialog





## [1.1.4](https://github.com/oscarmarina/blockquote-web-components/compare/@blockquote-web-components/blockquote-dialog@1.1.3...@blockquote-web-components/blockquote-dialog@1.1.4) (2024-07-25)

**Note:** Version bump only for package @blockquote-web-components/blockquote-dialog





## [1.1.3](https://github.com/oscarmarina/blockquote-web-components/compare/@blockquote-web-components/blockquote-dialog@1.1.2...@blockquote-web-components/blockquote-dialog@1.1.3) (2024-06-29)

**Note:** Version bump only for package @blockquote-web-components/blockquote-dialog





## [1.1.2](https://github.com/oscarmarina/blockquote-web-components/compare/@blockquote-web-components/blockquote-dialog@1.1.1...@blockquote-web-components/blockquote-dialog@1.1.2) (2024-06-23)

**Note:** Version bump only for package @blockquote-web-components/blockquote-dialog





## [1.1.1](https://github.com/oscarmarina/blockquote-web-components/compare/@blockquote-web-components/blockquote-dialog@1.1.0...@blockquote-web-components/blockquote-dialog@1.1.1) (2024-06-21)

**Note:** Version bump only for package @blockquote-web-components/blockquote-dialog





# [1.1.0](https://github.com/oscarmarina/blockquote-web-components/compare/@blockquote-web-components/blockquote-dialog@1.0.1...@blockquote-web-components/blockquote-dialog@1.1.0) (2024-06-09)

### Features

- **blockquote-dialog:** add labelledbyVisible property ([4effecd](https://github.com/oscarmarina/blockquote-web-components/commit/4effecd9ef68badca14504f935e11b9f25865534))

## [1.0.1](https://github.com/oscarmarina/blockquote-web-components/compare/@blockquote-web-components/blockquote-dialog@1.0.0...@blockquote-web-components/blockquote-dialog@1.0.1) (2024-06-07)

### Bug Fixes

- **blockquote-dialog:** update dependency ([d47ae21](https://github.com/oscarmarina/blockquote-web-components/commit/d47ae21abf13566a73fda7fd14fd09742d0d3eb9))

# 1.0.0 (2024-06-06)

### Features

- **blockquote-dialog:** new element ([20d977e](https://github.com/oscarmarina/blockquote-web-components/commit/20d977ef4132df1f0074cee09ebb12b4b0e6e670))
