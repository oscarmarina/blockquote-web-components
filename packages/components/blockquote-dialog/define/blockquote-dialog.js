import { BlockquoteDialog } from '../src/BlockquoteDialog.js';

window.customElements.define('blockquote-dialog', BlockquoteDialog);
