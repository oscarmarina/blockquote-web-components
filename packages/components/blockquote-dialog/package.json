{
  "name": "@blockquote-web-components/blockquote-dialog",
  "version": "1.4.7",
  "description": "Webcomponent blockquote-dialog following open-wc recommendations",
  "keywords": [
    "lit",
    "web-component",
    "lit-element"
  ],
  "license": "MIT",
  "author": "blockquote-dialog",
  "type": "module",
  "exports": {
    ".": {
      "default": "./src/index.js"
    },
    "./package.json": {
      "default": "./package.json"
    },
    "./BlockquoteDialog.js": {
      "default": "./src/BlockquoteDialog.js"
    },
    "./blockquote-dialog.js": {
      "default": "./define/blockquote-dialog.js"
    },
    "./index.js": {
      "default": "./src/index.js"
    },
    "./blockquote-dialog-styles.css.js": {
      "default": "./src/styles/blockquote-dialog-styles.css.js"
    }
  },
  "main": "src/index.js",
  "module": "src/index.js",
  "files": [
    "/src/",
    "!/**/*.scss"
  ],
  "scripts": {
    "analyze": "cem analyze --litelement --globs \"{src,define}/**/*.{js,ts}\" \"index.js\"",
    "build": "echo \"This is not a TypeScript project, so no need to build.\"",
    "dev:vite": "vite build",
    "format": "npm run format:eslint && npm run format:prettier && npm run format:stylelint",
    "format:eslint": "eslint . --fix",
    "format:prettier": "prettier \"**/*.{js,ts,css,scss,html,json}\" --write --ignore-path .prettierignore",
    "format:stylelint": "stylelint \"{demo,src}/**/*.{js,ts,css,scss,html}\" --fix --ignore-path .prettierignore",
    "postinstall": "npm run sort:package",
    "preview:vite": "vite preview",
    "sass:watch": "sass-style-template",
    "sort:package": "npx sort-package-json",
    "start": "concurrently -k -r \"npm:vite\" \"npm:sass:watch\"",
    "test": "vitest --run --browser.headless",
    "test:watch": "vitest",
    "vite": "vite"
  },
  "lint-staged": {
    "**/*.{js,ts,html}": [
      "npm run format"
    ],
    "**/*.json": [
      "npm run format:prettier"
    ],
    "**/*.{scss,css}": [
      "npm run format:prettier",
      "npm run format:stylelint"
    ]
  },
  "prettier": {
    "bracketSameLine": true,
    "bracketSpacing": false,
    "htmlWhitespaceSensitivity": "ignore",
    "printWidth": 100,
    "singleQuote": true,
    "trailingComma": "es5",
    "overrides": [
      {
        "files": "*.{scss,css}",
        "options": {
          "singleQuote": false
        }
      },
      {
        "files": "*.json",
        "options": {
          "parser": "json-stringify"
        }
      }
    ]
  },
  "stylelint": {
    "extends": "stylelint-config-standard-scss",
    "allowEmptyInput": true,
    "rules": {
      "custom-property-pattern": null,
      "no-duplicate-selectors": null,
      "color-function-notation": null,
      "alpha-value-notation": null,
      "media-feature-name-no-unknown": null,
      "comment-empty-line-before": null,
      "declaration-empty-line-before": null,
      "property-no-vendor-prefix": null,
      "selector-attribute-quotes": null,
      "value-no-vendor-prefix": null
    },
    "overrides": [
      {
        "files": [
          "**/*.html"
        ],
        "customSyntax": "postcss-html"
      },
      {
        "files": [
          "**/*.{ts,js}"
        ],
        "customSyntax": "postcss-lit"
      }
    ]
  },
  "dependencies": {
    "@blockquote-web-components/blockquote-directive-ariaidref-slot": "^1.3.7",
    "@blockquote/frontend-utilities": "^1.1.0",
    "lit": "^3.3.1"
  },
  "devDependencies": {
    "@blockquote-web-components/blockquote-base-common-dev-dependencies": "^1.12.3",
    "@blockquote-web-components/blockquote-base-embedded-webview": "^1.13.7",
    "@blockquote-web-components/blockquote-foundations-sass": "^1.1.7"
  },
  "publishConfig": {
    "access": "public"
  },
  "customElements": "custom-elements.json"
}
