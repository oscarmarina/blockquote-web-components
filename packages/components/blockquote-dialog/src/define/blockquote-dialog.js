import {BlockquoteDialog} from '../BlockquoteDialog.js';

window.customElements.define('blockquote-dialog', BlockquoteDialog);
