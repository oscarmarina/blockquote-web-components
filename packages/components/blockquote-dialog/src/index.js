export {BlockquoteDialog} from './BlockquoteDialog.js';
