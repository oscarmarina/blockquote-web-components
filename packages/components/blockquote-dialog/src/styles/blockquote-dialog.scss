:host {
  --_background-color: var(--blockquote-dialog-background-color, rgb(255, 255, 255));
  --_max-height: var(--blockquote-dialog-max-height, min(35rem, calc(100% - 3rem)));
  --_max-width: var(--blockquote-dialog-max-width, min(35rem, calc(100% - 3rem)));
  --_min-height: var(--blockquote-dialog-min-height, 8.75rem);
  --_min-width: var(--blockquote-dialog-min-width, 17.5rem);
  --_padding: var(--blockquote-padding, 1rem);

  box-sizing: border-box;
  display: contents;
  background-color: var(--_background-color);
  margin: auto;
  max-height: var(--_max-height);
  max-width: var(--_max-width);
  min-height: var(--_min-height);
  min-width: var(--_min-width);
  position: fixed;
  height: fit-content;
  width: fit-content;
}

:host([hidden]),
[hidden] {
  display: none !important;
}

*,
*::before,
*::after {
  box-sizing: inherit;
}

dialog {
  background: inherit;
  border: none;
  border-radius: inherit;
  flex-direction: column;
  margin: inherit;
  height: inherit;
  width: inherit;
  max-height: inherit;
  max-width: inherit;
  min-height: inherit;
  min-width: inherit;
  outline: none;
  overflow: visible;
  padding: 0;
}

:host([open]) dialog {
  display: flex;
}

.scroller {
  overflow-y: auto;
  min-height: var(--_min-height);
}

.content {
  padding: var(--_padding);
  min-height: inherit;
}
