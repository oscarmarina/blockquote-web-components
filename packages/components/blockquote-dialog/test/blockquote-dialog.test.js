/* eslint-disable lit-a11y/no-autofocus */
/* eslint-disable import/no-extraneous-dependencies */
import { html, fixture, assert, expect, fixtureCleanup, aTimeout } from '@open-wc/testing';
import sinon from 'sinon';
import '../define/blockquote-dialog.js';

const _formTpl = html`
  <form method="dialog">
    <header>
      <h2>Dialog header</h2>
    </header>
    <div>Dialog content</div>
    <footer>
      <menu>
        <button id="cancel" value="cancel">Cancel</button>
        <button id="confirm" value="confirm">Confirm</button>
      </menu>
    </footer>
  </form>
`;

const _formNoDialogTpl = html`
  <form>
    <header>
      <h2>Dialog header</h2>
    </header>
    <div>Dialog content</div>
    <footer>
      <menu>
        <button value="cancel">Cancel</button>
        <button value="confirm">Confirm</button>
      </menu>
    </footer>
  </form>
`;

const _formAutofocusTpl = html`
  <form>
    <header>
      <h2>Dialog header</h2>
      <button value="lose">Close</button>
    </header>
    <div>Dialog content</div>
    <div><input autofocus type="text" name="name" /></div>
    <footer>
      <menu>
        <button value="confirm">Confirm</button>
      </menu>
    </footer>
  </form>
`;

suite('BlockquoteDialog', () => {
  /**
   * @type {import('../src/index').BlockquoteDialog}
   */
  let el;

  teardown(() => fixtureCleanup());

  suite('Default Close', () => {
    setup(async () => {
      el = await fixture(html`
        <blockquote-dialog>${_formTpl}</blockquote-dialog>
      `);
    });

    suite('Semantic Dom and a11y', () => {
      test('SHADOW DOM - Structure test', async () => {
        await expect(el).shadowDom.to.equalSnapshot({ ignoreAttributes: ['id', 'name'] });
      });

      test('LIGHT DOM - Structure test', async () => {
        await expect(el).lightDom.to.equalSnapshot({ ignoreAttributes: ['id', 'name'] });
      });

      test('a11y', async () => {
        await assert.isAccessible(el);
      });
    });

    suite('Method and Property', () => {
      test('Method and Property Close', async () => {
        el.close();
        await el.updateComplete;
        assert.isNotTrue(el?.hasAttribute('open'));
      });
    });
  });

  suite('Default Close with "label"', () => {
    setup(async () => {
      el = await fixture(html`
        <blockquote-dialog label="aria-label Name">${_formTpl}</blockquote-dialog>
      `);
    });

    suite('Semantic Dom and a11y', () => {
      test('SHADOW DOM - Structure test', async () => {
        await expect(el).shadowDom.to.equalSnapshot({ ignoreAttributes: ['id', 'name'] });
      });

      test('LIGHT DOM - Structure test', async () => {
        await expect(el).lightDom.to.equalSnapshot({ ignoreAttributes: ['id', 'name'] });
      });

      test('a11y', async () => {
        await assert.isAccessible(el);
      });
    });

    suite('Method and Property', () => {
      test('Method and Property Close', async () => {
        el.close();
        await el.updateComplete;
        assert.isNotTrue(el?.hasAttribute('open'));
      });
    });
  });

  suite('Default Close role dialog', () => {
    setup(async () => {
      el = await fixture(html`
        <blockquote-dialog type="">${_formTpl}</blockquote-dialog>
      `);
    });

    suite('Semantic Dom and a11y', () => {
      test('SHADOW DOM - Structure test', async () => {
        await expect(el).shadowDom.to.equalSnapshot({ ignoreAttributes: ['id', 'name'] });
      });

      test('LIGHT DOM - Structure test', async () => {
        await expect(el).lightDom.to.equalSnapshot({ ignoreAttributes: ['id', 'name'] });
      });

      test('a11y', async () => {
        await assert.isAccessible(el);
      });
    });

    suite('Method and Property', () => {
      test('Method and Property Close', async () => {
        el.close();
        await el.updateComplete;
        assert.isNotTrue(el?.hasAttribute('open'));
      });
    });
  });

  suite('Default Open', () => {
    setup(async () => {
      el = await fixture(html`
        <blockquote-dialog labelledby="labelledby" open>
          ${_formTpl}
          <span slot="labelledby">All fields are required</span>
        </blockquote-dialog>
      `);
    });

    suite('Semantic Dom and a11y', () => {
      test('SHADOW DOM - Structure test', async () => {
        await expect(el).shadowDom.to.equalSnapshot({ ignoreAttributes: ['id', 'name'] });
      });

      test('LIGHT DOM - Structure test', async () => {
        await expect(el).lightDom.to.equalSnapshot({ ignoreAttributes: ['id', 'name'] });
      });

      test('a11y', async () => {
        await assert.isAccessible(el);
      });
    });

    suite('Method and Property', () => {
      test('Method and Property Open', async () => {
        el.show();
        await el.updateComplete;
        assert.isTrue(el?.hasAttribute('open'));
      });
    });
  });

  suite('Event', () => {
    setup(async () => {
      el = await fixture(html`
        <blockquote-dialog open label="aria-label Name">${_formTpl}</blockquote-dialog>
      `);
    });

    test('Dialog should close when "Click" in backdrop', async () => {
      const dialog = el.shadowRoot?.querySelector('dialog');
      dialog?.click();
      await el.updateComplete;
      assert.isNotTrue(el?.hasAttribute('open'));
    });

    test('Dialog should not close when clicking inside the content', async () => {
      const content = el.shadowRoot?.querySelector('.content');
      /** @type {HTMLElement} */ (content)?.click();
      await el.updateComplete;
      assert.isTrue(el?.hasAttribute('open'));
    });

    test('The "cancel" event should close the dialog if the target is the same', async () => {
      const dialog = el.shadowRoot?.querySelector('dialog');
      const event = new Event('cancel');
      dialog?.dispatchEvent(event);
      await el.updateComplete;
      assert.isNotTrue(el?.hasAttribute('open'));
    });

    test('The "cancel" event should not close the dialog if the target is different from the dialog itself', async () => {
      const content = el.shadowRoot?.querySelector('.content');
      const event = new Event('cancel', { bubbles: true });
      content?.dispatchEvent(event);
      await el.updateComplete;
      assert.isTrue(el?.hasAttribute('open'));
    });

    test('The "close" event should close the dialog', async () => {
      const dialog = el.shadowRoot?.querySelector('dialog');
      const event = new Event('close');
      dialog?.dispatchEvent(event);
      await el.updateComplete;
      assert.isNotTrue(el?.hasAttribute('open'));
    });

    test('The "close" set returnValue', async () => {
      const button = el?.querySelector('button');
      button?.click();
      await el.updateComplete;
      await aTimeout(200);
      const { returnValue } = el;
      assert.equal(returnValue.tagName, 'BUTTON');
    });
  });

  suite('Form - with method', () => {
    setup(async () => {
      el = await fixture(html`
        <blockquote-dialog label="aria-label Name">${_formTpl}</blockquote-dialog>
      `);
    });

    test('The dialog should close when the form has the dialog method', async () => {
      const button = el?.querySelector('button');
      button?.click();
      await el.updateComplete;
      assert.isNotTrue(el?.hasAttribute('open'));
    });
  });

  suite('Form - without method', () => {
    setup(async () => {
      el = await fixture(html`
        <blockquote-dialog open label="aria-label Name">${_formNoDialogTpl}</blockquote-dialog>
      `);
    });

    test('Dialog does not should close when "Form Submit"', async () => {
      const button = el?.querySelector('button');
      const form = el?.querySelector('form');
      form?.addEventListener('submit', ev => ev.preventDefault());
      button?.click();
      await el.updateComplete;
      assert.isTrue(el?.hasAttribute('open'));
    });
  });

  suite('focus', () => {
    setup(async () => {
      el = await fixture(html`
        <blockquote-dialog label="aria-label Name">${_formTpl}</blockquote-dialog>
      `);
    });

    test('focus on first focus trap sets focus to last element', async () => {
      el.open = true;
      await el.updateComplete;
      const lastButton = el.querySelector('#confirm');
      const spy = sinon.spy(/** @type {HTMLButtonElement} */ (lastButton), 'focus');
      assert.isTrue(spy.notCalled);
      await el.updateComplete;
      const firstFocusTrap = el.shadowRoot?.querySelector('dialog > span:first-of-type');
      /** @type {HTMLElement} */ (firstFocusTrap)?.focus();
      assert.isTrue(spy.called);
    });

    test('focus on first focus trap sets focus to first element when relatedTarget is undefined', async () => {
      el.open = true;
      await el.updateComplete;
      const firstButton = el.querySelector('#cancel');
      const spy = sinon.spy(/** @type {HTMLButtonElement} */ (firstButton), 'focus');
      assert.isTrue(spy.notCalled);
      await el.updateComplete;
      const firstFocusTrap = el.shadowRoot?.querySelector('dialog > span:first-of-type');
      /** @type {HTMLElement} */ (firstFocusTrap)?.dispatchEvent(
        new FocusEvent('focus', { relatedTarget: null }),
      );
      assert.isTrue(spy.called);
    });

    test('focus on last focus trap sets focus to first element', async () => {
      el.open = true;
      await el.updateComplete;
      const firstButton = el.querySelector('#cancel');
      const spy = sinon.spy(/** @type {HTMLButtonElement} */ (firstButton), 'focus');
      assert.isTrue(spy.notCalled);
      await el.updateComplete;
      const lastFocusTrap = el.shadowRoot?.querySelector('dialog > span:last-of-type');
      /** @type {HTMLElement} */ (lastFocusTrap)?.focus();
      assert.isTrue(spy.called);
    });

    test('focus on last focus trap sets focus to last element when relatedTarget is undefined', async () => {
      el.open = true;
      await el.updateComplete;
      const lastButton = el.querySelector('#confirm');
      const spy = sinon.spy(/** @type {HTMLButtonElement} */ (lastButton), 'focus');
      assert.isTrue(spy.notCalled);
      await el.updateComplete;
      const firstFocusTrap = el.shadowRoot?.querySelector('dialog > span:last-of-type');
      /** @type {HTMLElement} */ (firstFocusTrap)?.dispatchEvent(
        new FocusEvent('focus', { relatedTarget: null }),
      );
      assert.isTrue(spy.called);
    });

    test('autofocus', async () => {
      /**
       * @type {import('../src/index').BlockquoteDialog}
       */
      const elAutofocus = await fixture(html`
        <blockquote-dialog label="aria-label Name">${_formAutofocusTpl}</blockquote-dialog>
      `);

      elAutofocus.open = true;
      await elAutofocus.updateComplete;
      assert.isTrue(document.activeElement?.tagName === 'INPUT');
    });
  });
});
