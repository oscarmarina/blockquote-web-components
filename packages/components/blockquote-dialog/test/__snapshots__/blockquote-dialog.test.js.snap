// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BlockquoteDialog > Default Close > Semantic Dom and a11y > LIGHT DOM - Structure test > LIGHT DOM 1`] = `
"<blockquote-dialog role="presentation">
  <form method="dialog">
    <header>
      <h2>
        Dialog header
      </h2>
    </header>
    <div>
      Dialog content
    </div>
    <footer>
      <menu>
        <button value="cancel">
          Cancel
        </button>
        <button value="confirm">
          Confirm
        </button>
      </menu>
    </footer>
  </form>
</blockquote-dialog>
"
`;

exports[`BlockquoteDialog > Default Close > Semantic Dom and a11y > SHADOW DOM - Structure test > SHADOW DOM 1`] = `
"<dialog role="alertdialog">
  <span
    hidden=""
    tabindex="0"
  >
  </span>
  <div class="scroller">
    <div class="content">
      <slot>
      </slot>
    </div>
  </div>
  <span
    hidden=""
    tabindex="0"
  >
  </span>
</dialog>
"
`;

exports[`BlockquoteDialog > Default Close role dialog > Semantic Dom and a11y > LIGHT DOM - Structure test > LIGHT DOM 1`] = `
"<blockquote-dialog
  role="presentation"
  type=""
>
  <form method="dialog">
    <header>
      <h2>
        Dialog header
      </h2>
    </header>
    <div>
      Dialog content
    </div>
    <footer>
      <menu>
        <button value="cancel">
          Cancel
        </button>
        <button value="confirm">
          Confirm
        </button>
      </menu>
    </footer>
  </form>
</blockquote-dialog>
"
`;

exports[`BlockquoteDialog > Default Close role dialog > Semantic Dom and a11y > SHADOW DOM - Structure test > SHADOW DOM 1`] = `
"<dialog role="alertdialog">
  <span
    hidden=""
    tabindex="0"
  >
  </span>
  <div class="scroller">
    <div class="content">
      <slot>
      </slot>
    </div>
  </div>
  <span
    hidden=""
    tabindex="0"
  >
  </span>
</dialog>
"
`;

exports[`BlockquoteDialog > Default Close with "label" > Semantic Dom and a11y > LIGHT DOM - Structure test > LIGHT DOM 1`] = `
"<blockquote-dialog
  label="aria-label Name"
  role="presentation"
>
  <form method="dialog">
    <header>
      <h2>
        Dialog header
      </h2>
    </header>
    <div>
      Dialog content
    </div>
    <footer>
      <menu>
        <button value="cancel">
          Cancel
        </button>
        <button value="confirm">
          Confirm
        </button>
      </menu>
    </footer>
  </form>
</blockquote-dialog>
"
`;

exports[`BlockquoteDialog > Default Close with "label" > Semantic Dom and a11y > SHADOW DOM - Structure test > SHADOW DOM 1`] = `
"<dialog role="alertdialog">
  <span
    hidden=""
    tabindex="0"
  >
  </span>
  <div class="scroller">
    <div class="content">
      <slot>
      </slot>
    </div>
  </div>
  <span
    hidden=""
    tabindex="0"
  >
  </span>
</dialog>
"
`;

exports[`BlockquoteDialog > Default Open > Semantic Dom and a11y > LIGHT DOM - Structure test > LIGHT DOM 1`] = `
"<blockquote-dialog
  labelledby="labelledby"
  open=""
  role="presentation"
>
  <form method="dialog">
    <header>
      <h2>
        Dialog header
      </h2>
    </header>
    <div>
      Dialog content
    </div>
    <footer>
      <menu>
        <button value="cancel">
          Cancel
        </button>
        <button value="confirm">
          Confirm
        </button>
      </menu>
    </footer>
  </form>
  <span slot="labelledby">
    All fields are required
  </span>
</blockquote-dialog>
"
`;

exports[`BlockquoteDialog > Default Open > Semantic Dom and a11y > SHADOW DOM - Structure test > SHADOW DOM 1`] = `
"<dialog role="alertdialog">
  <span
    hidden=""
    tabindex="0"
  >
  </span>
  <div class="scroller">
    <div class="content">
      <slot>
      </slot>
    </div>
  </div>
  <span
    hidden=""
    tabindex="0"
  >
  </span>
</dialog>
"
`;
